File: bot.js
/**
 * እሁድን በፍቅር ዲጂታል ፕሮ v5.7.4 - Group Menu & Station Expert
 * ቴክኖሎጂ፡ Telegraf (Telegram Bot API), sqlite (Database), Node.js
 * ማሻሻያ፡ በግሩፕ ውስጥ እንደ ሜኑ አዝራር የሚያገለግል የ'Pay Here' ጣቢያ እና የ'/setup' መመሪያ ተጨምሯል
 */

require('dotenv').config();
const { Telegraf, session, Markup } = require('telegraf');
const sqlite3 = require('sqlite3');
const { open } = require('sqlite');
const http = require('http');

// --- 1. HEALTH CHECK SERVER ---
const PORT = process.env.PORT || 3000;
http.createServer((req, res) => {
    res.writeHead(200, { 'Content-Type': 'text/plain; charset=utf-8' });
    res.end('EdirPay Enterprise Bot is Operational');
}).listen(PORT, '0.0.0.0');

// --- 2. CONFIGURATION ---
const BOT_TOKEN = process.env.BOT_TOKEN;
const ADMIN_ID = process.env.ADMIN_ID ? parseInt(process.env.ADMIN_ID) : null;
const FINANCE_ID = process.env.FINANCE_ID ? parseInt(process.env.FINANCE_ID) : null;
const TEST_GROUP_ID = process.env.TEST_GROUP_ID ? parseInt(process.env.TEST_GROUP_ID) : null;

const MINI_APP_URL = process.env.MINI_APP_URL || "https://abiym.github.io/edirpaybot/";
const DB_FILE = process.env.DISK_PATH ? `${process.env.DISK_PATH}/edir_pro_final.db` : 'edir_pro_final.db';

if (!BOT_TOKEN || !ADMIN_ID) {
    console.error("❌ CRITICAL: BOT_TOKEN or ADMIN_ID is missing!");
    process.exit(1);
}

const bot = new Telegraf(BOT_TOKEN);
let db;

// --- 3. DATABASE INITIALIZATION ---
async function initDB() {
    db = await open({
        filename: DB_FILE,
        driver: sqlite3.Database
    });

    await db.exec(`
        CREATE TABLE IF NOT EXISTS members (
            user_id INTEGER PRIMARY KEY, 
            username TEXT, 
            full_name TEXT, 
            balance REAL DEFAULT 0,
            joined_at TEXT
        );
        CREATE TABLE IF NOT EXISTS payments (
            id INTEGER PRIMARY KEY AUTOINCREMENT, 
            tx_id TEXT UNIQUE, 
            user_id INTEGER, 
            username TEXT, 
            purpose TEXT, 
            period TEXT, 
            amount REAL, 
            penalty REAL DEFAULT 0, 
            pay_for TEXT,
            gateway TEXT,
            file_id TEXT, 
            status TEXT DEFAULT 'AWAITING_PHOTO', 
            group_msg_id INTEGER, 
            timestamp TEXT,
            reviewer_name TEXT,
            treasurer_name TEXT
        );
    `);
    console.log("✅ Database Ready.");
}

// --- 4. HELPERS & LOGGING ---
bot.use(session());

bot.catch((err, ctx) => {
    console.error(`❌ Bot Logic Error [${ctx.updateType}]:`, err);
});

bot.use(async (ctx, next) => {
    if (ctx.message && ctx.message.text) {
        console.log(`[INCOMING] ${ctx.chat.type}: ${ctx.message.text} from ${ctx.from.id}`);
    }
    return next();
});

const isReviewer = (id) => id === ADMIN_ID;
const isTreasurer = (id) => id === FINANCE_ID || (id === ADMIN_ID && !FINANCE_ID);
const generateTXID = () => `#EUDE${Math.floor(1000 + Math.random() * 9000)}`;

const formatPaymentReport = (p, emoji, statusText) => {
    let base = `======= የክፍያ ሪፖርት ${p.tx_id} =======\n\n` +
               `👤 አባል: @${p.username}\n` +
               `🎯 ዓላማ: ${p.purpose}\n` +
               `💰 መጠን: ${p.amount} ብር\n` +
               `⚠️ ቅጣት: ${p.penalty || 0} ብር\n` +
               `💳 መንገድ: ${p.gateway || 'MANUAL'}\n` +
               `📅 ጊዜ: ${p.period}\n` +
               `━━━━━━━━━━━━━━━━━━\n` +
               `${emoji} ሁኔታ: ${statusText}`;
    if (p.reviewer_name) base += `\n🔍 መርማሪ: ${p.reviewer_name}`;
    if (p.treasurer_name) base += `\n🏦 ፋይናንስ: ${p.treasurer_name}`;
    return base;
};

// --- 5. INLINE MODE HANDLER ---

bot.on('inline_query', async (ctx) => {
    const results = [
        {
            type: 'article',
            id: 'pay_now',
            title: '💳 ክፍያ ይፈጽሙ (Pay Now)',
            description: 'የእሁድን በፍቅር ሚኒ አፕ በመጠቀም ክፍያ ለመፈጸም እዚህ ይጫኑ',
            input_message_content: { message_text: `ሰላም! የዕድር ክፍያ ለመፈጸም ከታች ያለውን አዝራር ይጠቀሙ።` },
            reply_markup: { inline_keyboard: [[{ text: "📱 ሚኒ አፑን ክፈት", web_app: { url: MINI_APP_URL } }]] },
            thumb_url: 'https://cdn-icons-png.flaticon.com/512/10149/10149458.png'
        }
    ];
    return await ctx.answerInlineQuery(results, { cache_time: 0 });
});

// --- 6. BOT COMMANDS ---

// ቦቱ ወደ ግሩፕ ሲገባ የሚላክ ሰላምታ
bot.on('new_chat_members', async (ctx) => {
    const isBotAdded = ctx.message.new_chat_members.some(m => m.id === ctx.botInfo.id);
    if (isBotAdded) {
        const text = `ሰላም! 🖐 እኔ **የእሁድን በፍቅር** ዲጂታል ዕድር ቦት ነኝ።\n\n` +
                     `አባላት በዚህ ግሩፕ ውስጥ ሆነው ክፍያ ለመፈጸም እንዲችሉ እባክዎ **Admin** ያድርጉኝ።\n\n` +
                     `ለመጀመር፦ /setup ብለው ይላኩ።`;
        return ctx.replyWithMarkdown(text);
    }
});

bot.start(async (ctx) => {
    try {
        const time = new Date().toLocaleString('am-ET');
        await db.run(
            'INSERT OR IGNORE INTO members (user_id, username, full_name, joined_at) VALUES (?, ?, ?, ?)',
            ctx.from.id, ctx.from.username || 'N/A', ctx.from.first_name, time
        );
        const text = `ሰላም ${ctx.from.first_name}! 👋 ወደ **እሁድን በፍቅር** ዲጂታል ዕድር እንኳን ደህና መጡ።`;
        return await ctx.replyWithMarkdown(text, {
            reply_markup: { inline_keyboard: [[{ text: "📱 መተግበሪያውን ክፈት", web_app: { url: MINI_APP_URL } }]] }
        });
    } catch (e) { console.error("Start Error:", e); }
});

// [SETUP COMMAND] - ለግሩፕ አድሚኖች መመሪያ
bot.command('setup', async (ctx) => {
    const text = `⚙️ **የግሩፕ አቀነባበር መመሪያ**\n━━━━━━━━━━━━━━━━━━\n` +
                 `ቦቱ በግሩፕ ውስጥ እንደ "ሜኑ አዝራር" ሆኖ እንዲያገለግል እነዚህን 2 ደረጃዎች ይከተሉ፦\n\n` +
                 `1️⃣ **አስተዳዳሪ ያድርጉኝ፦** ቦቱ መልዕክቶችን የማንበብ እና የማጥፋት መብት ያለው አድሚን መሆን አለበት።\n\n` +
                 `2️⃣ **ጣቢያውን ይፍጠሩ፦** በግሩፕ ውስጥ \`/payhere\` ብለው ይላኩ። የሚመጣውን መልዕክት **Pin** በማድረግ አባላት ሁልጊዜ እንዲያገኙት ያድርጉ።\n\n` +
                 `3️⃣ **Command Menu፦** በ @BotFather በኩል 'Edit Commands' በማድረግ 'pay' እና 'payhere' የሚሉትን ይጨምሩ።`;
    return ctx.replyWithMarkdown(text);
});

// [PAYHERE STATION] - የግሩፑ 'ሜኑ አዝራር'
bot.command('payhere', async (ctx) => {
    const stationText = `🏦 **የእሁድን በፍቅር ዲጂታል ዕድር ጣቢያ**\n` +
                        `━━━━━━━━━━━━━━━━━━\n` +
                        `ሰላም አባላት! 👋 በዚህ ግሩፕ ውስጥ ክፍያ ለመፈጸም ወይም ቁጠባዎን ለማየት ከታች ያለውን አዝራር ይጠቀሙ።\n\n` +
                        `💡 **ማሳሰቢያ፦** መረጃውን ከላኩ በኋላ ደረሰኝዎን ለቦቱ (@${ctx.botInfo.username}) በግል መላክዎን አይርሱ።\n` +
                        `━━━━━━━━━━━━━━━━━━`;
    try {
        await ctx.replyWithMarkdown(stationText, {
            reply_markup: { inline_keyboard: [[{ text: "💳 ክፍያ ይፈጽሙ / ሚኒ አፕ", web_app: { url: MINI_APP_URL } }]] }
        });
        return ctx.reply("☝️ **የክፍያ ጣቢያው ተፈጥሯል። አባላት በቀላሉ እንዲያገኙት እባክዎ ይህንን መልዕክት ፒን (Pin) ያድርጉት።**");
    } catch (e) {
        console.error("PayHere Error:", e);
        return ctx.reply("❌ አዝራሩን መላክ አልተቻለም። ቦቱ አድሚን መሆኑን ያረጋግጡ።");
    }
});

bot.command('pay', async (ctx) => {
    try {
        return await ctx.reply(`ሰላም ${ctx.from.first_name}! ክፍያ ለመፈጸም ከታች ያለውን አዝራር ይጫኑ፦`, {
            reply_markup: { inline_keyboard: [[{ text: "💳 ክፍያ ይፈጽሙ", web_app: { url: MINI_APP_URL } }]] }
        });
    } catch (e) {
        console.error("Pay Error:", e);
        return ctx.reply("❌ እባክዎ ቦቱን በግል ቻት ያነጋግሩ።");
    }
});

bot.command('admin', async (ctx) => {
    if (!isReviewer(ctx.from.id) && !isTreasurer(ctx.from.id)) return;
    try {
        const stats = await db.get("SELECT SUM(balance) as total, COUNT(*) as count FROM members");
        const pStats = await db.get("SELECT SUM(penalty) as tp FROM payments WHERE status = 'APPROVED'");
        const pending = await db.get("SELECT COUNT(*) as count FROM payments WHERE status IN ('PENDING_REVIEW', 'PENDING_TREASURY')");
        await ctx.replyWithMarkdown(`📊 **የአስተዳዳሪ ማጠቃለያ**\n━━━━━━━━━━━━━━━━━━\n👥 አባላት: ${stats.count}\n💰 ቁጠባ: ${stats.total || 0} ብር\n⚠️ ቅጣት: ${pStats.tp || 0} ብር\n⏳ ጥያቄዎች: ${pending.count}`);
    } catch (e) { console.error("Admin Error:", e); }
});

bot.command('id', async (ctx) => ctx.reply(`Chat ID: \`${ctx.chat.id}\``, { parse_mode: 'Markdown' }));

// --- 7. DATA & MEDIA HANDLING ---

bot.on('web_app_data', async (ctx) => {
    try {
        const data = JSON.parse(ctx.webAppData.data.json());
        if (data.type === 'payment_report') {
            const txId = generateTXID();
            const timestamp = new Date().toLocaleString('am-ET');
            const gateway = data.gateway ? data.gateway.toUpperCase() : "MANUAL";
            await db.run("DELETE FROM payments WHERE user_id = ? AND status = 'AWAITING_PHOTO'", ctx.from.id);
            await db.run(
                `INSERT INTO payments (tx_id, user_id, username, purpose, period, amount, penalty, pay_for, gateway, timestamp) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
                txId, ctx.from.id, ctx.from.username || ctx.from.first_name, data.purpose, data.period, data.amount, data.penalty, data.payFor, gateway, timestamp
            );
            await ctx.replyWithMarkdown(`✅ የ${data.amount} ብር መረጃ ተመዝግቧል፦ \`${txId}\`\n\n📷 **አሁን ደረሰኝዎን (Screenshot) ይላኩ።**`);
        } 
    } catch (err) { console.error("WebAppData Error:", err); }
});

bot.on(['photo', 'document'], async (ctx) => {
    try {
        const pending = await db.get("SELECT * FROM payments WHERE user_id = ? AND status = 'AWAITING_PHOTO' ORDER BY id DESC LIMIT 1", ctx.from.id);
        if (!pending) return;
        const fileId = ctx.message.photo ? ctx.message.photo.pop().file_id : ctx.message.document.file_id;
        await db.run("UPDATE payments SET status = 'PENDING_REVIEW', file_id = ? WHERE id = ?", fileId, pending.id);
        if (TEST_GROUP_ID) {
            const report = formatPaymentReport(pending, "⏳", "በምርመራ ላይ...");
            const sent = await bot.telegram.sendMessage(TEST_GROUP_ID, report);
            await db.run('UPDATE payments SET group_msg_id = ? WHERE id = ?', sent.message_id, pending.id);
        }
        const revKb = Markup.inlineKeyboard([[Markup.button.callback("✅ አጽድቅ", `rev_app_${pending.id}`), Markup.button.callback("❌ ውድቅ", `rev_rej_${pending.id}`)]]);
        await bot.telegram.sendPhoto(ADMIN_ID, fileId, { caption: `🚨 ምርመራ: ${pending.tx_id}`, ...revKb });
        await ctx.reply(`📩 ደረሰኝዎ ለምርመራ ደርሷል። ሲጸድቅ እናሳውቆታለን።`);
    } catch (err) { console.error("Upload Error:", err); }
});

// --- 8. APPROVAL ACTIONS ---

bot.action(/^(rev_app|rev_rej|tr_app|tr_rej)_(\d+)$/, async (ctx) => {
    try {
        const [tier, action, id] = ctx.match;
        const p = await db.get("SELECT * FROM payments WHERE id = ?", id);
        if (!p) return ctx.answerCbQuery("Not found");
        const adminName = ctx.from.first_name;

        if (tier === 'rev') {
            if (!isReviewer(ctx.from.id)) return ctx.answerCbQuery("Denied");
            if (action === 'app') {
                await db.run("UPDATE payments SET status = 'PENDING_TREASURY', reviewer_name = ? WHERE id = ?", adminName, id);
                const trKb = Markup.inlineKeyboard([[Markup.button.callback("✅ ባንክ ገብቷል", `tr_app_${id}`), Markup.button.callback("❌ ውድቅ", `tr_rej_${id}`)]]);
                await bot.telegram.sendPhoto(FINANCE_ID || ADMIN_ID, p.file_id, { caption: `🏦 ባንክ ማረጋገጫ: ${p.tx_id}\n✅ Reviewed by ${adminName}`, ...trKb });
                if (TEST_GROUP_ID && p.group_msg_id) {
                    const up = await db.get("SELECT * FROM payments WHERE id = ?", id);
                    await bot.telegram.editMessageText(TEST_GROUP_ID, p.group_msg_id, null, formatPaymentReport(up, "⏳", "ደረሰኝ ተረጋግጧል፤ ባንክ እየታየ ነው..."));
                }
            } else {
                await db.run("UPDATE payments SET status = 'REJECTED' WHERE id = ?", id);
                await bot.telegram.sendMessage(p.user_id, `❌ ክፍያዎ ውድቅ ተደርጓል (#${p.tx_id})`);
            }
        }
        if (tier === 'tr') {
            if (!isTreasurer(ctx.from.id)) return ctx.answerCbQuery("Finance only");
            if (action === 'app') {
                await db.run("UPDATE payments SET status = 'APPROVED', treasurer_name = ? WHERE id = ?", adminName, id);
                const total = (p.amount || 0) + (p.penalty || 0);
                await db.run("UPDATE members SET balance = balance + ? WHERE user_id = ?", total, p.user_id);
                await bot.telegram.sendMessage(p.user_id, `✅ ክፍያዎ ጽድቋል (#${p.tx_id})\nአጠቃላይ ብር፦ ${total}`);
                if (TEST_GROUP_ID && p.group_msg_id) {
                    const final = await db.get("SELECT * FROM payments WHERE id = ?", id);
                    await bot.telegram.editMessageText(TEST_GROUP_ID, p.group_msg_id, null, formatPaymentReport(final, "✅", "ተረጋግጦ ጽድቋል"));
                }
            } else {
                await db.run("UPDATE payments SET status = 'REJECTED' WHERE id = ?", id);
                await bot.telegram.sendMessage(p.user_id, `❌ ክፍያዎ ውድቅ ተደርጓል (#${p.tx_id})`);
            }
        }
        await ctx.editMessageCaption(`${ctx.callbackQuery.message.caption}\n🏁 ውሳኔ: ${action} በ: ${adminName}`);
        ctx.answerCbQuery("Done");
    } catch (e) { console.error("Action Error:", e); }
});

// --- 9. STARTUP & CONFLICT RESOLUTION ---
const sleep = (ms) => new Promise(r => setTimeout(r, ms));

async function startBot(retries = 10) {
    try {
        await initDB();
        console.log("🧹 Webhook ማጽዳት...");
        await bot.telegram.deleteWebhook({ drop_pending_updates: true });
        
        // በግሩፕ ውስጥ የሚታዩ ትዕዛዞችን ማዘጋጀት
        await bot.telegram.setMyCommands([
            { command: 'pay', description: 'ክፍያ ለመጀመር' },
            { command: 'payhere', description: 'የክፍያ ጣቢያ በግሩፕ ለመፍጠር' },
            { command: 'setup', description: 'የግሩፕ አቀነባበር መመሪያ' },
            { command: 'admin', description: 'የአስተዳዳሪ ማጠቃለያ' }
        ]);

        console.log("⏳ ኮኔክሽን በማስተካከል ላይ...");
        await sleep(6000); 
        await bot.launch({ dropPendingUpdates: true });
        console.log("🚀 EdirPay Enterprise v5.7.4 Online!");
    } catch (err) {
        if (err.response && err.response.error_code === 409 && retries > 0) {
            console.warn(`⚠️ Conflict. Retrying in 10s... (${retries} left)`);
            await sleep(10000);
            return startBot(retries - 1);
        }
        console.error("❌ Fatal Startup Error:", err);
    }
}
startBot();

process.once('SIGINT', () => bot.stop('SIGINT'));
process.once('SIGTERM', () => bot.stop('SIGTERM'));

File: bot.py
import asyncio
import json
import logging
import os
import sqlite3
from datetime import datetime
from aiohttp import web
from aiogram import Bot, Dispatcher, types, F
from aiogram.filters import Command
from aiogram.utils.keyboard import ReplyKeyboardBuilder, InlineKeyboardBuilder
from aiogram.types import WebAppInfo
from dotenv import load_dotenv

# --- 1. CONFIGURATION & LOGGING ---
load_dotenv()
logging.basicConfig(level=logging.INFO)

BOT_TOKEN = os.getenv("BOT_TOKEN")
# አስተዳዳሪዎችን ለመለየት (ለምሳሌ: 1062635928)
admin_env = os.getenv("ADMIN_IDS", "1062635928")
ADMIN_IDS = [int(id.strip()) for id in admin_env.replace(',', ' ').split()]
MINI_APP_URL = os.getenv("MINI_APP_URL")
TEST_GROUP_ID = os.getenv("TEST_GROUP_ID")
if TEST_GROUP_ID:
    TEST_GROUP_ID = int(TEST_GROUP_ID)

DB_FILE = 'edir_pro_final.db'

if not BOT_TOKEN:
    raise ValueError("❌ BOT_TOKEN missing in .env file!")

bot = Bot(token=BOT_TOKEN)
dp = Dispatcher()

# --- 2. DATABASE SETUP ---
def init_db():
    conn = sqlite3.connect(DB_FILE)
    cursor = conn.cursor()
    cursor.execute('''
        CREATE TABLE IF NOT EXISTS members (
            user_id INTEGER PRIMARY KEY, 
            username TEXT, 
            full_name TEXT, 
            tier TEXT DEFAULT 'መሠረታዊ', 
            total_savings REAL DEFAULT 0, 
            joined_at TEXT
        )
    ''')
    cursor.execute('''
        CREATE TABLE IF NOT EXISTS payments (
            id INTEGER PRIMARY KEY AUTOINCREMENT, 
            user_id INTEGER, 
            username TEXT, 
            gateway TEXT, 
            purpose TEXT, 
            period TEXT, 
            total_amount REAL, 
            penalty REAL DEFAULT 0, 
            pay_for_member TEXT, 
            file_id TEXT, 
            status TEXT DEFAULT 'AWAIT_APPROVAL', 
            timestamp TEXT
        )
    ''')
    cursor.execute('''
        CREATE TABLE IF NOT EXISTS notifications (
            id INTEGER PRIMARY KEY AUTOINCREMENT, 
            user_id INTEGER, 
            title TEXT, 
            message TEXT, 
            type TEXT, 
            timestamp TEXT
        )
    ''')
    conn.commit()
    conn.close()

# --- 3. RENDER STABILITY (HTTP SERVER) ---
async def handle_ping(request):
    return web.Response(text="EdirPay Bot (Python) is Active")

async def start_http_server():
    app = web.Application()
    app.router.add_get('/', handle_ping)
    runner = web.AppRunner(app)
    await runner.setup()
    site = web.TCPSite(runner, '0.0.0.0', int(os.getenv("PORT", 3000)))
    await site.start()

# --- 4. BOT HANDLERS ---

def is_admin(user_id):
    return user_id in ADMIN_IDS

@dp.message(Command("start"))
async def cmd_start(message: types.Message):
    now = datetime.now().strftime("%d/%m/%Y %H:%M:%S")
    conn = sqlite3.connect(DB_FILE)
    cursor = conn.cursor()
    cursor.execute('INSERT OR IGNORE INTO members (user_id, username, full_name, joined_at) VALUES (?, ?, ?, ?)',
                   (message.from_user.id, message.from_user.username or 'N/A', message.from_user.first_name, now))
    conn.commit()
    conn.close()

    builder = ReplyKeyboardBuilder()
    builder.row(types.KeyboardButton(text="📱 ሚኒ አፑን ክፈት", web_app=WebAppInfo(url=MINI_APP_URL)))
    
    if is_admin(message.from_user.id):
        builder.row(types.KeyboardButton(text="⚙️ የአስተዳዳሪ ሁነታ (Admin Mode)"))
    
    builder.row(types.KeyboardButton(text="📊 ሁኔታዬን እይ"), types.KeyboardButton(text="❓ እርዳታ"))
    
    await message.reply(
        f"እንኳን ወደ **እሁድን በፍቅር** መጡ! 👋 v25.2.2 (Python)\n\nእባክዎ ሚኒ አፑን በመክፈት የክፍያ ሪፖርትዎን ይላኩ።",
        reply_markup=builder.as_markup(resize_keyboard=True),
        parse_mode="Markdown"
    )

@dp.message(Command("id"))
async def cmd_id(message: types.Message):
    await message.reply(
        f"📌 **የቻት መረጃ**\n\n👤 ስም: *{message.chat.title or 'የግል ቻት'}*\n"
        f"🆔 ID: `{message.chat.id}`\n"
        f"🌐 ዓይነት: `{message.chat.type}`\n\n"
        f"💡 ይህንን ID በሬንደር ላይ በ `TEST_GROUP_ID` ቦታ ያስገቡት።",
        parse_mode="Markdown"
    )

@dp.message(F.content_type == types.ContentType.WEB_APP_DATA)
async def handle_webapp_data(message: types.Message):
    try:
        data = json.loads(message.web_app_data.data)
        if data.get('type') == 'payment_report':
            time_now = datetime.now().strftime("%d/%m/%Y %H:%M:%S")
            
            # ለግሩፕ ማሳወቂያ መላክ
            if TEST_GROUP_ID:
                try:
                    await bot.send_message(
                        TEST_GROUP_ID,
                        f"🔔 **አዲስ የክፍያ ሪፖርት ደርሷል**\n\n"
                        f"👤 አባል: @{message.from_user.username or message.from_user.first_name}\n"
                        f"🎯 ዓላማ: {data['purpose']}\n"
                        f"💰 መጠን: {data['amount']} ብር\n"
                        f"💳 መንገድ: {data['gateway'].upper()}\n\n"
                        f"✅ አስተዳዳሪዎች እባካችሁ በግል ገብታችሁ አጽድቁ።",
                        parse_mode="Markdown"
                    )
                except Exception as e:
                    logging.error(f"Group notification error: {e}")

            if data['gateway'] == 'manual':
                # ሴሽን ለመምሰል እዚህ ጋር ዳታውን ለጊዜው እናስቀምጣለን (ቀላል በሆነ መንገድ)
                # በእውነተኛ ስራ ላይ Redis ወይም FSM መጠቀም ይመከራል
                await message.answer(f"✅ የ{data['amount']} ብር መረጃ ተመዝግቧል። 📷 እባክዎ ደረሰኝዎን አሁን ይላኩ።")
                # ለቀላልነት ዳታውን በዴታቤዝ ውስጥ 'PENDING_FILE' ብለን እናስቀምጠዋለን
                conn = sqlite3.connect(DB_FILE)
                cursor = conn.cursor()
                cursor.execute('''
                    INSERT INTO payments (user_id, username, gateway, purpose, period, total_amount, penalty, pay_for_member, status, timestamp)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                ''', (message.from_user.id, message.from_user.username or 'N/A', data['gateway'], data['purpose'], data['period'], data['amount'], data['penalty'], data.get('payFor', 'self'), 'AWAITING_FILE', time_now))
                conn.commit()
                conn.close()
            else:
                conn = sqlite3.connect(DB_FILE)
                cursor = conn.cursor()
                cursor.execute('''
                    INSERT INTO payments (user_id, username, gateway, purpose, period, total_amount, penalty, pay_for_member, timestamp)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
                ''', (message.from_user.id, message.from_user.username or 'N/A', data['gateway'], data['purpose'], data['period'], data['amount'], data['penalty'], data.get('payFor', 'self'), time_now))
                conn.commit()
                conn.close()
                await message.answer(f"🚀 የ{data['gateway'].upper()} ክፍያዎ ተመዝግቧል። ሲረጋገጥ እናሳውቆታለን።")
    except Exception as e:
        logging.error(f"WebAppData Error: {e}")

@dp.message(F.photo | F.document)
async def handle_receipt_upload(message: types.Message):
    file_id = message.photo[-1].file_id if message.photo else message.document.file_id
    
    conn = sqlite3.connect(DB_FILE)
    cursor = conn.cursor()
    # ፋይል የሚጠባበቅ የመጨረሻ ክፍያ መፈለግ
    cursor.execute('SELECT id FROM payments WHERE user_id = ? AND status = "AWAITING_FILE" ORDER BY id DESC LIMIT 1', (message.from_user.id,))
    row = cursor.fetchone()
    
    if row:
        cursor.execute('UPDATE payments SET file_id = ?, status = "AWAIT_APPROVAL" WHERE id = ?', (file_id, row[0]))
        conn.commit()
        await message.answer("📩 ደረሰኝዎ ለፋይናንስ ኦፊሰር ተልኳል። እናመሰግናለን!")
    conn.close()

# --- 5. ADMIN ACTIONS ---

@dp.message(F.text == "⚙️ የአስተዳዳሪ ሁነታ (Admin Mode)")
async def admin_mode(message: types.Message):
    if not is_admin(message.from_user.id): return
    
    conn = sqlite3.connect(DB_FILE)
    cursor = conn.cursor()
    cursor.execute('SELECT COUNT(*) FROM payments WHERE status = "AWAIT_APPROVAL"')
    pending_count = cursor.fetchone()[0]
    conn.close()

    builder = ReplyKeyboardBuilder()
    builder.row(types.KeyboardButton(text=f"📑 የሚጠባበቁ ክፍያዎች ({pending_count})"))
    builder.row(types.KeyboardButton(text="📈 አጠቃላይ ሪፖርት"), types.KeyboardButton(text="👤 ወደ አባልነት ተመለስ"))
    
    await message.answer("🛠 የአስተዳዳሪ መቆጣጠሪያ ማዕከል", reply_markup=builder.as_markup(resize_keyboard=True))

@dp.callback_query(F.data.startswith("app_") | F.data.startswith("rej_"))
async def process_approval(callback: types.CallbackQuery):
    if not is_admin(callback.from_user.id):
        await callback.answer("ፈቃድ የለዎትም!")
        return

    action, pay_id = callback.data.split("_")
    new_status = "APPROVED" if action == "app" else "REJECTED"
    
    conn = sqlite3.connect(DB_FILE)
    cursor = conn.cursor()
    cursor.execute('SELECT user_id, total_amount FROM payments WHERE id = ?', (pay_id,))
    payment = cursor.fetchone()
    
    if payment:
        cursor.execute('UPDATE payments SET status = ? WHERE id = ?', (new_status, pay_id))
        if action == "app":
            cursor.execute('UPDATE members SET total_savings = total_savings + ? WHERE user_id = ?', (payment[1], payment[0]))
            try:
                await bot.send_message(payment[0], f"✅ የ{payment[1]} ብር ክፍያዎ ተረጋግጦ ጽድቋል።")
            except: pass
        conn.commit()
    conn.close()

    await callback.message.edit_caption(caption=f"{callback.message.caption}\n\n🏁 **ውሳኔ:** {'✅ ጸድቋል' if action == 'app' else '❌ ውድቅ ተደርጓል'}")
    await callback.answer("ተጠናቋል")

# --- 6. MAIN ---
async def main():
    init_db()
    # የሬንደር ሰርቨርን እና ቦቱን በአንድ ላይ ማስነሳት
    await asyncio.gather(
        start_http_server(),
        dp.start_polling(bot)
    )

if __name__ == "__main__":
    asyncio.run(main())
